docs/design/topics.md

```diff
@@ -0,0 +1,168 @@
+[Topics][hg-topics]. There also is the [Topic feature][gerrit-topics] in Gerrit,
+which groups commits with a single identifier.
+
+
+## Goals and non-goals
+
+### Goals
+
+The goals for this Project are small, see below.
+
+* Introduce the concept of native topological branches for Jujutsu.
+* Simplify Git interop by reducing the burden on Bookmarks.
+* Add Change metadata as a storage concept.
+* Remove the awkward `bookmark` to Git `branch` mapping.
+
+### Non-Goals
+
+* Making bookmarks unnecessary.
+
+## Overview
+
+Until now, Jujutsu had no native set of topological branches, just 
+[Bookmarks][bm] which interact poorly with Git's expectation of branches. 
+Topics on the otherhand are can be made to represent Git branches as users 
+expect them, see [Julia Evans poll][jvns-poll]. They also allow us to 
+seamlessly take over the [tracking-branches][tb] concept.
+
+Other use-cases they're useful for are representing a set of 
+[archived commits][archived] or even a [checkout history][checkout].
+
+
+#### Behavior
+
+A Topic is a set of Changes marked with a name, which infectiously moves 
+forward with each descendant you create. All Changes without a named topic are
+implicitly in a anonymous topic, which gets separately tracked as soon as you 
+send it in for review (so materialize it as a Git branch). A Topic may have
+multiple heads. A single Change may belong to multiple Topics, which means 
+that they also can get exported as overlapping Git branches. 
+
+
+TODO: Example here
+
+#### Major Use-Cases 
+
+
+##### Obsoletion of `[experimental.auto-advance-bookmarks]`
+
+Since Topics are infectious by nature, they can perfectly map to a Git branch.
+This alleviates the need for the auto-advancing bookmarks. It also makes the 
+use-case of the solo developer working on Git `main` easier, as you just mark
+all descendants of `trunk()` belonging to the `main` topic, which then gets 
+translated to the `main` branch on `jj git push`.
+
+
+##### Change Archival
+
+For the archival use-case the infectious property of a Topic isn't as 
+important. So having a 
+`jj topic create -r 'trunk()..@ & description("archive:")' should mark them all
+revisions with a `archive:` description as archived. For this use-case the 
+non-contiguous property of Topics really shine.
+
+##### Checkout history
+
+### Detailed Design
+
+#### Git Interop
+
+For all continuous Topics we can just simply export them as Git Branches. For 
+Topics which contain non-contiguous parts, we should allow exporting them 
+one-by-one or with a generated name, such as `<topic-name>-1` for each part.
+
+#### Storage
+
+We should store `Topics` as metadata on the serialized proto, without 
+considering the resulting Gencode. 
+
+
+```protobuf
+// A simple Key-Value pair. 
+message StringPair {
+  string key = 1;
+  string value = 2;
+  // Could be extended by a protobuf.Any see the future possibilities section.
+}
+
+message Commit {
+  //...
+  repeated StringPair metadata = N;
```

Comment:
For context, Google would probably use this for shipping things like "submitted_piper_change_number: 123" from the server to the client